endpoints/openrtb2/auction.go

```diff
@@ -341,6 +370,97 @@ func setSiteImplicitly(httpReq *http.Request, bidReq *openrtb.BidRequest) {
+	// Parse the Stored Request IDs from the BidRequest and Imps.
+	storedBidRequestId, hasStoredBidRequest, err := getStoredRequestId(requestJson)
+	if err != nil {
+		return nil, []error{err}
+	}
+	imps, impIds, idIndices, errs := parseImpInfo(requestJson)
+	if len(errs) > 0 {
+		return nil, errs
+	}
+
+	// Fetch all of the Stored Request data
+	var allIds = make([]string, len(impIds), len(impIds)+1)
+	copy(allIds, impIds)
+	if hasStoredBidRequest {
+		allIds = append(allIds, storedBidRequestId)
+	}
+	storedRequests, errs := deps.storedReqFetcher.FetchRequests(ctx, allIds)
+	if len(errs) > 0 {
+		return nil, errs
+	}
+
+	// Apply the Stored BidRequest, if it exists
+	resolvedRequest := requestJson
+	if hasStoredBidRequest {
+		resolvedRequest, err = jsonpatch.MergePatch(storedRequests[storedBidRequestId], requestJson)
+		if err != nil {
+			return nil, []error{err}
+		}
+	}
+
+	// Apply any Stored Imps, if they exist. Since the JSON Merge Patch overrides arrays,
+	// and Prebid Server defers to the HTTP Request to resolve conflicts, it's safe to
+	// assume that the request.imp data did not change when applying the Stored BidRequest.
+	for i := 0; i < len(impIds); i++ {
+		resolvedImp, err := jsonpatch.MergePatch(storedRequests[impIds[i]], imps[i])
+		if err != nil {
+			return nil, []error{err}
+		}
+		imps[idIndices[i]] = resolvedImp
+	}
+	if len(impIds) > 0 {
+		newImpJson, err := json.Marshal(imps)
+		if err != nil {
+			return nil, []error{err}
+		}
+		resolvedRequest, err = jsonparser.Set(resolvedRequest, newImpJson, "imp")
+		if err != nil {
+			return nil, []error{err}
+		}
+	}
+
+	return resolvedRequest, nil
+}
+
+func parseImpInfo(requestJson []byte) (imps []json.RawMessage, ids []string, impIdIndices []int, errs []error) {
+	if impArray, dataType, _, err := jsonparser.Get(requestJson, "imp"); err == nil && dataType == jsonparser.Array {
+		i := 0
+		jsonparser.ArrayEach(impArray, func(imp []byte, dataType jsonparser.ValueType, offset int, err error) {
+			if storedImpId, hasStoredImp, err := getStoredRequestId(imp); err != nil {
+				errs = append(errs, err)
+			} else if hasStoredImp {
+				imps = append(imps, imp)
+				ids = append(ids, storedImpId)
+				impIdIndices = append(impIdIndices, i)
+			}
+			i++
+		})
+	}
+	return
+}
+
+// getStoredRequestId parses a Stored Request ID from some json, without doing a full (slow) unmarshal.
+// It returns the ID, true/false whether a stored request key existed, and an error if anything went wrong
+// (e.g. malformed json, id not a string, etc).
+func getStoredRequestId(data []byte) (string, bool, error) {
+	// These keys must be kept in sync with openrtb_ext.ExtStoredRequest
+	value, dataType, _, err := jsonparser.Get(data, "ext", "prebid", "storedrequest", "id")
+	if dataType == jsonparser.NotExist {
+		return "", false, nil
+	}
+	if err != nil {
+		return "", false, err
+	}
+	if dataType != jsonparser.String {
+		return "", true, errors.New("ext.prebid.storedrequest.id must be a string")
```

Comment:
Returning true here means if there is a value, but not a string, we add "" to the list of IDs we need to look up, and then we will try to merge whatever IdFetcher returns for that ID (I assume ""). This all should be a noop, but an expensive one. This could potentially get very weird in edge cases. I think we should just return false for non-strings.

Reply:
are you sure? If you're right, I should definitely fix this, but... the two calls are:

```
if storedImpId, hasStoredImp, err := getStoredRequestId(imp); err != nil {
  errs = append(errs, err)
} else if hasStoredImp {
  imps = append(imps, imp)
  ids = append(ids, storedImpId)
  impIdIndices = append(impIdIndices, i)
}
```

and

```
storedBidRequestId, hasStoredBidRequest, err := getStoredRequestId(requestJson)
if err != nil {
  return nil, []error{err}
}
```

Both check the error condition before true/false... so they should result in an immediate 400 response.

One of the new test cases is supposed to validate this as well: https://github.com/prebid/prebid-server/pull/273/files#diff-f77b4bc57e914a8b0878c96f85b8783aR640

Reply:
Oh, yes you are correct, the error checks will keep it off of the list of imps to process.